app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,168 @@
-  rescue ActiveRecord::RecordNotFound => exception
-    info "DeleteObjectHierarchyWorker##{method_name} raised #{exception.class} with message #{exception.message}"
+  before_perform do |job|
+    info "Starting #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
   end
 
-  protected
+  after_perform do |job|
+    if @remaining_hierarchy.present?
+      self.class.perform_later(*@remaining_hierarchy)
+      msg = " iteration of"
+    end
+    info "Finished#{msg} #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
+  end
 
-  delegate :info, to: 'Rails.logger'
+  class << self
+    # convenience method to schedule deleting an active record object
+    def delete_later(ar_object)
+      perform_later *hierarchy_entries_for(ar_object)
+    end
 
-  attr_reader :object, :caller_worker_hierarchy
+    private
 
-  def build_batch
-    batch = Sidekiq::Batch.new
-    batch.description = batch_description
-    batch_callbacks(batch) { batch.jobs { destroy_and_delete_associations } }
-    batch
-  end
+    # @return the hierarchy entries to handle deletion of an object
+    def hierarchy_entries_for(ar_object)
+      return [] unless ar_object.persisted? # e.g. when calling Proxy#oidc_configuration a new object can be generated
 
-  def batch_description
-    "Deleting #{object.class.name} [##{object.id}]"
-  end
+      if ar_object.is_a?(Account) && !ar_object.should_be_deleted?
+        raise DoNotRetryError, "background deleting account #{ar_object.id} which is not scheduled for deletion"
+      end
 
-  def batch_callbacks(batch)
-    %i[success complete].each { |name| batch.on(name, self.class, callback_options) }
-    yield
-    bid = batch.bid
+      ar_object_str = "#{ar_object.class}-#{ar_object.id}"
 
-    if Sidekiq::Batch::Status.new(bid).total.zero?
-      on_complete(bid, callback_options)
-    else
-      info("DeleteObjectHierarchyWorker#batch_success_callback retry job with the hierarchy of workers: #{caller_worker_hierarchy}")
-      retry_job wait: 5.minutes
+      [
+        "Plain-#{ar_object_str}",
+        *ar_object.background_deletion.map { "Association-#{ar_object_str}:#{_1}" },
+      ]
     end
   end
 
-  def destroy_and_delete_associations
-    Array(object.background_deletion).each do |association_config|
-      reflection = BackgroundDeletion::Reflection.new(association_config)
-      next unless destroyable_association?(reflection.name)
+  # @param hierarchy [Array<String>] something like ["Plain-Service-1234", "Association-Service-1234:plans" ...]
+  # @note processes deleting a hierarchy of objects and reschedules itself at current progress after a time limit
+  def perform(*hierarchy)
+    return compatibility(hierarchy) unless hierarchy.first.is_a?(String)
 
-      ReflectionDestroyer.new(object, reflection, caller_worker_hierarchy).destroy_later
+    started = now
+    while hierarchy.present? && now - started < WORK_TIME_LIMIT_SECONDS
```

Comment:
So, to make sure I understand correctly:
we process the array of hierarchy for N seconds (5 at the moment), and when the time is up we reschedule whatever is still left?

Reply:
Correct